src/compiler/transformers/classFields.ts

```diff
@@ -189,6 +189,51 @@ namespace ts {
             }
         }
 
+        function visitorThisOrSuperInStaticFieldContext(root: Expression | undefined, receiver: LeftHandSideExpression, baseClass: LeftHandSideExpression | undefined, hasClassDecorators: boolean | undefined): Expression | undefined {
+            const shouldRewriteSuper = baseClass && languageVersion >= ScriptTarget.ES2015;
+            const shouldAvoidThisOrSuper = hasClassDecorators;
+            const shouldCareAboutSuper = shouldRewriteSuper || shouldAvoidThisOrSuper;
+
+            function staticFieldThisOrSuperVisitor(node: Expression): Expression {
+                if (!(node.transformFlags & (TransformFlags.ContainsLexicalThis | (shouldCareAboutSuper ? TransformFlags.ContainsLexicalSuper : TransformFlags.None)))) {
+                    return node;

```

Comment:
I would add a comment that this will be visited by the caller to `visitorThisOrSuperInStaticFieldContext`. I would note that I'd generally prefer a single walk of the tree, as opposed to visiting a subtree one way and then visiting it again a second time.